algorithms/gr-dfsrh/typescript/depth_first_search.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Tree node representation
+ *
+ * @class TreeNode
+ */
+class TreeNode {
+    label: number;
+    data: number;
+  
+    /**
+     * Creates an instance of TreeNode.
+     * @param {number} label
+     * @param {number} data
+     */
+    constructor(label: number, data: number){
+        this.label = label;
+        this.data = data;
+    }
+  }
+  
+  /**
+   * Tree representation
+   *
+   * @class Tree
+   */
+  class Tree {
+    // The root of a the tree
+    root: TreeNode
+    // We use adjacency list to for BFS/DFS, we keep only the labels here
+    // not the actual nodes
+    adjacencies: Array<Array<number>>
+    // Array of the nodes
+    nodes: Array<TreeNode>
+    // Allow a node to be connected with itself, cyclic graph
+    allowCyclic: boolean
+  
+    constructor(rootValue: number, allowCyclic: boolean = false) {
+      this.root = new TreeNode(0, rootValue);
+      this.adjacencies = [];
+      this.adjacencies[0] = [];
+      this.nodes = [];
+      this.allowCyclic = allowCyclic;
+    }
+  
+    /**
+     * Add a node to the tree
+     *
+     * @param {TreeNode} node
+     */
+    addNode(node: TreeNode){
+      this.adjacencies[node.label] = [];
+      this.nodes.push(node);
+    }
+  
+    /**
+     * Set a connection between two nodes
+     *
+     * @param {number} node1
+     * @param {number} node2
+     */
+    setConnection(node1Label: number, node2Label: number){
+      if(!this.allowCyclic && node1Label === node2Label) {
+        throw Error("Graph cannot be cyclic");
+      }
+      this.adjacencies[node1Label].push(node2Label);
+    }
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
```

Comment:
Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)